dask_ndmeasure/__init__.py

```diff
@@ -70,6 +72,93 @@ def center_of_mass(input, labels=None, index=None):
     return com_lbl
 
 
+def labeled_comprehension(input,
+                          labels,
+                          index,
+                          func,
+                          out_dtype,
+                          default,
+                          pass_positions=False):
+    """
+    Roughly equivalent to [func(input[labels == i]) for i in index].
+
+    Sequentially applies an arbitrary function (that works on array_like input)
+    to subsets of an n-D image array specified by `labels` and `index`.
+    The option exists to provide the function with positional parameters as the
+    second argument.
+
+    Parameters
+    ----------
+    input : array_like
+        Data from which to select `labels` to process.
+    labels : array_like or None
+        Labels to objects in `input`.
+        If not None, array must be same shape as `input`.
+        If None, `func` is applied to raveled `input`.
+    index : int, sequence of ints or None
+        Subset of `labels` to which to apply `func`.
+        If a scalar, a single value is returned.
+        If None, `func` is applied to all non-zero values of `labels`.
+    func : callable
+        Python function to apply to `labels` from `input`.
+    out_dtype : dtype
+        Dtype to use for `result`.
+    default : int, float or None
+        Default return value when a element of `index` does not exist
+        in `labels`.
+    pass_positions : bool, optional
+        If True, pass linear indices to `func` as a second argument.
+        Default is False.
+
+    Returns
+    -------
+    result : ndarray
+        Result of applying `func` to each of `labels` to `input` in `index`.
+    """
+
+    input, labels, index = _utils._norm_input_labels_index(
+        input, labels, index
+    )
+    out_dtype = numpy.dtype(out_dtype)
+    default = numpy.array(default, dtype=out_dtype)[()]
+    pass_positions = bool(pass_positions)
+
+    indices = _utils._ravel_shape_indices(
+        input.shape, dtype=numpy.int64, chunks=input.chunks
+    )
+
+    lbl_mtch = _utils._get_label_matches(labels, index)
+
+    lbl_mtch_any = lbl_mtch.any(
+        axis=tuple(_pycompat.irange(index.ndim, lbl_mtch.ndim))
+    )
+
+    result = numpy.empty(index.shape, dtype=object)
+    for i in itertools.product(*[_pycompat.irange(j) for j in index.shape]):
+        args = (input[lbl_mtch[i]],)
+        if pass_positions:
+            args += (indices[lbl_mtch[i]],)
+
+        result[i] = dask.delayed(_utils._labeled_comprehension_func)(
+            func, out_dtype, default, lbl_mtch_any[i], *args
+        )
+        result[i] = dask.array.from_delayed(result[i], tuple(), out_dtype)
```

Comment:
It seems that we are unable to get Dask to conditionally evaluate the user provided function, `func`, based on whether the computation needs to happen or not. Thus we use a thunk-like strategy to wrap the call to `func` with `_labeled_comprehension_func` wherein we can perform the check. As a result, we can ensure the user provided function is only called as needed.

Though it would be nice if we could avoid this workaround. By using `delayed` we effectively force the entire computation per object to take place on one worker in memory. If the objects get to be too large to fit in memory, this could present an issue. Fixing this will require some conditional evaluation strategy as mentioned above. This will require some changes in Dask to support conditional evaluation. Have raised issue ( https://github.com/dask/dask/issues/2579 ) on this point.